### PR TITLE
Add `integration:matrix` and `integration:single` mage targets

### LINKED ISSUE #2714: [Integration] Add `mage integration:test -matrix` option to expand tests to run on all supported OS's
**Describe the enhancement:**

At the moment the `mage integration:test` only runs on the smallest set of instances it can create. This is great for quick testing but doesn't provide a full test of the Elastic Agent on all supported OS's.

Adding a `-matrix` flag to the `mage integration:test` target would allow the `testing/runner.Runner` to expand the instances spawned to be all the different OS types that the runner supports. As long as the test is not specific on its definition then it can run on X number of hosts.

Example lets say a test says it can only run on Windows:

```
define.Require(define.Requirements{
    OS: []define.OS{Type: define.Windows},
}
```

Then with the `-matrix` option that test would be ran on `Windows Server 2012`, `Windows Server 2016`, `Windows Server 2019`.

**Describe a specific use case for the enhancement or feature:**

To provide broad coverage of integration and end-to-end testing of the Elastic Agent.

### LINKED ISSUE #2716: [Integration] Add `mage integration:test -run [TestName]` to target a specific test
# Issue

**Describe the enhancement:**

Currently calling `mage integration:test` runs all tests on remote hosts. Add a new `-run [TestName]` option to the `integration:test` target to allow the the `testing/runner` to only deploy a single instance to run that one specific test.

I think a couple of other options could be useful:

`-os [os_type]` - Target a specific OS, in the case it supports multiple
`-arch [arch_type]` - Target a specific architecture, in the case it supports multiple
`-distro [distro]` - Target a specific distro, in the case it supports multiple
`-version [version]` - Target a specific version of the distro, in the case it supports multiple

**Describe a specific use case for the enhancement or feature:**

To help developers to write and debug tests.

# Definition of done

- [ ] Add a way to run only one specific test instead of every tests
- [ ] Add some options to choose the specific OS on which the test should be run

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds a new `integration:matrix` target that explodes the possible instances to run tests on to include all possible combinations. `integration:test` always just picks the first for each OS type, the `integration:matrix` will run it on the same OS type if the runner supports different distros and versions of that type. Currently running `integration:test` the tests only run on Ubuntu 22.04, running `integration:matrix` the tests are run on Ubuntu 22.04 and Ubuntu 20.04.

Adds a new `integration:single [testName]` target that allows the selection of only a single integration test to run. This makes it easier on the developer to work and fix a single test without having to run all tests.

Fixes issue where the mage target would not return a bad exit code in the case of test failures. This now will exit 1 if any tests fails, exit 0 if all are successful.

## Why is it important?

`integration:matrix` is important so we can get complete test coverage across our entire support matrix.

`integration:single [testName` is needed for developers to write and fix a single test.

Exit code change is required by CI to allow the CI to know if the test run was successful.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Related issues

- Closes #2714 
- Closes #2716

### CODE REVIEW

**Review by ycombinator — COMMENTED** [review]
Code changes LGTM.  Please mention the new mage targets in https://github.com/elastic/elastic-agent/blob/main/docs/test-framework-dev-guide.md#running-tests.  Thanks.

**Comment by pierrehilbert** [comment]
Agree with @ycombinator here, please don't forget to update the documentation every time we are adding new things or changing existing one.

File: magefile.go [hunk 1]
```diff
@@ -1423,7 +1409,49 @@ func (Integration) TestOnRemote(ctx context.Context) error {
...
+	}
+	err = writeFile("build/TEST-go-integration.xml", results.XMLOutput, 0644)
+	if err != nil {
+		return err
+	}
+	if results.Failures > 0 {
+		fmt.Printf(">>> Testing completed (%d failures)\n", results.Failures)
+	} else {
+		fmt.Printf(">>> Testing completed (%d successful)\n", results.Tests)
+	}
```
Comment (AndersonQ) [inline magefile.go:1444]: [Suggestion]
Print both, the success and failures for the failures case.
```suggestion
	if results.Failures > 0 {
		fmt.Printf(">>> Testing completed (%d failures, %d successful)\n",
		 results.Failures, results.Tests-results.Failures)
	} else {
		fmt.Printf(">>> Testing completed (%d successful)\n", results.Tests)
	}
```

File: pkg/testing/runner/runner.go [hunk 2]
```diff
@@ -807,29 +838,96 @@ func findLayoutBatchByID(id string, batches []LayoutBatch) (LayoutBatch, bool) {
...
+		batches = append(batches, LayoutBatch{
+			ID:       xid.New().String(),
+			LayoutOS: s,
+			Batch:    batch,
+			Skip:     true,
+		})
+		return batches, nil
+	} else if err != nil {
+		return nil, err
+	}
```
Comment (AndersonQ) [inline pkg/testing/runner/runner.go:864]: [NIT]
Fail fast

```suggestion
	if err!=nil {
	if !errors.Is(err, ErrOSNotSupported) {
	    return nil, err
	}
		var s LayoutOS
		s.OS.Type = batch.OS.Type
		s.OS.Arch = batch.OS.Arch
		s.OS.Distro = batch.OS.Distro
		if s.OS.Distro == "" {
			s.OS.Distro = "unknown"
		}
		if s.OS.Version == "" {
			s.OS.Version = "unknown"
		}
		batches = append(batches, LayoutBatch{
			ID:       xid.New().String(),
			LayoutOS: s,
			Batch:    batch,
			Skip:     true,
		})
		return batches, nil
	}
```

File: pkg/testing/runner/runner.go [hunk 3]
```diff
@@ -807,29 +838,96 @@ func findLayoutBatchByID(id string, batches []LayoutBatch) (LayoutBatch, bool) {
 	return LayoutBatch{}, false
 }
 
-func createBatch(batch define.Batch) (LayoutBatch, error) {
-	skip := false
-	specific, err := getSupported(batch.OS)
+func createBatches(batch define.Batch, matrix bool) ([]LayoutBatch, error) {
+	specifics, err := getSupported(batch.OS)
+	var batches []LayoutBatch
```
Comment (AndersonQ) [inline pkg/testing/runner/runner.go:843]: [Go convention]
```suggestion
	var batches []LayoutBatch
	specifics, err := getSupported(batch.OS)
```

**Review by AndersonQ — APPROVED** [review]
Two questions:
 - why not making the `single` to take a regex, just like `go test -run`?
 - wouldn't it make sense to have a test filter for the `matrix` also?
I imagine a single test failing for an OS that isn't the one selected by `single`

besides just suggestions, fell free to ignore them if you like

**Review by ycombinator — APPROVED** [review]

**Pull request merged**